features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c

```diff
@@ -0,0 +1,333 @@
+                                    block_size - cmac_ctx->unprocessed_len : ilen;
+        memcpy( &cmac_ctx->unprocessed_block[cmac_ctx->unprocessed_len],
+                input, size_to_copy );
+        cmac_ctx->unprocessed_len += size_to_copy;
+        input += size_to_copy;
+        ilen -= size_to_copy;
+
+        /*
+         * Process the unproccessed data, in case it reached a full AES block,
+         * and there is there is still input data.
+         */
+        if( cmac_ctx->unprocessed_len == SASI_AES_BLOCK_SIZE_IN_BYTES && ilen > 0 )
+        {
+            if( SaSi_AesBlock( &cmac_ctx->CC_Context, cmac_ctx->unprocessed_block,
+                              SASI_AES_BLOCK_SIZE_IN_BYTES, NULL ) != 0 )
+            {
+                ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+                goto exit;
+            }
+            cmac_ctx->unprocessed_len = 0;
+        }
+    }
+
+    if( ilen > 0 )
+    {
+        const size_t size_to_store = ( ilen % SASI_AES_BLOCK_SIZE_IN_BYTES == 0) ?
+                                     SASI_AES_BLOCK_SIZE_IN_BYTES : ilen % SASI_AES_BLOCK_SIZE_IN_BYTES;
+        memcpy( &cmac_ctx->unprocessed_block[0],
+                           input + ilen - size_to_store,
+                           size_to_store );
+        cmac_ctx->unprocessed_len = size_to_store;
+        ilen -= size_to_store;
+        if( ilen > 0 )
+        {
+            if( SaSi_AesBlock( &cmac_ctx->CC_Context, ( uint8_t * )input,
+                               ilen, NULL ) != 0 )
+            {
+                ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+                goto exit;
+            }
+        }
+    }
+
+exit:
+    if( ret != 0 )
+    {
+        mbedtls_platform_zeroize( cmac_ctx, sizeof( *cmac_ctx ) );
+        mbedtls_free( cmac_ctx );
+        SaSi_AesFree( &cmac_ctx->CC_Context );
+    }
+    return( ret );
+}
+
+int mbedtls_cipher_cmac_finish( mbedtls_cipher_context_t *ctx,
+                                unsigned char *output )
+{
+    mbedtls_cmac_context_t* cmac_ctx;
+    int ret = 0;
+    size_t olen = SASI_AES_BLOCK_SIZE_IN_BYTES;
+
+    if( ctx == NULL || ctx->cipher_info == NULL ||
+        ctx->cmac_ctx == NULL || output == NULL )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    cmac_ctx = ctx->cmac_ctx;
+
+    if( cmac_ctx->is_cc_initiated == 0 )
+    {
+        ret = init_cc( cmac_ctx );
+        if( ret != 0 )
+            goto exit;
+    }
+
+    if( ( ret = SaSi_AesFinish( &cmac_ctx->CC_Context, cmac_ctx->unprocessed_len,
+                                cmac_ctx->unprocessed_block,
+                                cmac_ctx->unprocessed_len, output, &olen ) ) != 0 )
+    {
+        ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+        goto exit;
+    }
+
+exit:
+    if( SaSi_AesFree( &cmac_ctx->CC_Context ) != 0 && ret == 0 )
+    {
+        ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+    }
+
+    return( ret );
+}
+
+int mbedtls_cipher_cmac_reset( mbedtls_cipher_context_t *ctx )
```

Comment:
How is the sequence of events supposed to work if the user wants to reset the cmac? Is it something like:
```
starts
update
update 
...
finish
reset
starts
update
update
...
```

Or is it something like:
```
starts
update
update 
...
finish
reset
update
update
...
```
Does the second one even make sense? How would the user change the key (for example)? Also there would probably be a memory leak in that case because the cmac context will be reallocated.

Reply:
According to [documentation for `mbedtls_cmac_reset()`](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/cmac.h#L138):
~~~

 *                      It is called after mbedtls_cipher_cmac_finish()
 *                      and before mbedtls_cipher_cmac_update().
~~~
[mbedtls_cipher_cmac_finish()`](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/cmac.h#L119):
~~~
 *                      It is called after mbedtls_cipher_cmac_update().
 *                      It can be followed by mbedtls_cipher_cmac_reset() and
 *                      mbedtls_cipher_cmac_update(), or mbedtls_cipher_free().
~~~
[`mbedtls_cipher_cmac_update()`](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/cmac.h#L100):
~~~
*                      It is called between mbedtls_cipher_cmac_starts() or
 *                      mbedtls_cipher_cmac_reset(), and mbedtls_cipher_cmac_finish().
 *                      Can be called repeatedly.
~~~

Reply:
In other words, we do not need to call starts after reset.